src/domain/common/entities/ShareableEntity.ts

```diff
@@ -0,0 +1,37 @@
+import _ from "lodash";
+import { SharedRef } from "./Ref";
+import { User } from "../../user/entities/User";
+
+export class ShareableEntity<T extends SharedRef> {
+    constructor(protected data: T) {}
+
+    public hasPermissions(permission: "read" | "write", currentUser: User) {
+        if (currentUser.isGlobalAdmin) return true;
+
+        const {
+            publicAccess = "--------",
+            userAccesses = [],
+            userGroupAccesses = [],
+            user,
+        } = this.data;
+        const token = permission === "read" ? "r" : "w";
+
+        const isUserOwner = user.id === currentUser.id;
```

Comment:
I get a bug on some instances here, seems user can be undefined but TS does not complain. We might have some types wrong:

```ts
const isUserOwner = user?.id === currentUser.id;
```

Reply:
Ok, fixed!